tests/unit/tls_test.cc

```diff
@@ -268,36 +269,43 @@ class https_server {
     }
 };
 
+struct with_https_server {
+    std::unique_ptr<https_server> server;
+    void setup() {
+        server = std::make_unique<https_server>();
+    }
+    void teardown() {
+        server.reset();
+    }
+};
+
 #if !SEASTAR_TESTING_WITH_NETWORKING
 
-SEASTAR_THREAD_TEST_CASE(test_simple_x509_client) {
+SEASTAR_THREAD_FIXTURE_TEST_CASE(test_simple_x509_client, with_https_server) {
     auto certs = ::make_shared<tls::certificate_credentials>();
-    https_server server;
-    certs->set_x509_trust_file(server.cert(), tls::x509_crt_format::PEM).get();
-    connect_to_ssl_addr(certs, server.addr(), server.name()).get();
+    certs->set_x509_trust_file(server->cert(), tls::x509_crt_format::PEM).get();
+    connect_to_ssl_addr(certs, server->addr(), server->name()).get();
 }
```

Comment:
I understand the motivation, but the macro proliferation makes it harder to understand the role of the inputs to SEASTAR_THREAD_FIXURE_TEST_CASE.

If we stick to macro-less C++, we'd write `with_https_server_thread(std::function<void (https_server&)>). ` and use that in the test case. Isn't that easier to read?


/cc @nyh

Reply:
i see, you are following the route of `test_env::do_with_async()`.

the reason to change `tls_test.cc` is just to demonstrate how `SEASTAR_THREAD_FIXURE_TEST_CASE` works. but it might not be the best place for this. i can drop this change if it hurts the readability in this very case.

what i really want is to split https://github.com/scylladb/scylladb/blob/23be6f03360ba4943da55ac5128c7f6398264476/test/boost/sstable_compaction_test.cc#L939-L1007 into smaller tests. i think there are two places where we can improve the test:

1. a fixture which provides the `test_env`. we already have a `test_env` . despite that the way how we customize it is a little bit awkward -- to pass a 2nd parameter after the huge body of the test.  
2. a helper class which provides all the helper functions. these helper functions are now implemented as lambdas. 

by introducing these macros to Seastar, we should be able to use them in other tests as well. i understand that the fixture macro might hurt the readability, but this pattern is not invented here. but it depends on if developers are used to the fixture macros in other testing framework. AFAICT, it is already widely used by popular test frameworks like [GoogleTest](http://google.github.io/googletest/reference/testing.html#TEST_F) and [Boost.Test](https://live.boost.org/doc/libs/1_83_0/libs/test/doc/html/boost_test/tests_organization/fixtures/case.html). so as long as developers are familiar with these test frameworks, they can apply what they've been using when writing tests for Seastar applications.

Reply:
> what i really want is to split https://github.com/scylladb/scylladb/blob/23be6f03360ba4943da55ac5128c7f6398264476/test/boost/sstable_compaction_test.cc#L939-L1007 into smaller tests. i think there are two places where we can improve the test:
> 

i think we are talking about different things. could you please take a look my reply above?

Reply:
My point is that if you have specific goals in mind why you think fixtures are useful, please demonstrate **these** goals, and not a non-interesting example. You said it helps with test_env but I didn't understand how exactly. Is the main goal of the fixtures to reduce one (or several) layers of indentation with ugly lambda? If that's the case, you should show an example that does this (your example didn't).
The "helper functions" thing doesn't make sense to me. C++ already has plenty of features to do this - you can have file-static functions (this old-school approach is underappreciated today, I don't know why). You can have classes. You can instantiate classes with RAII. I still don't understand how fixtures help for that.

In general, my main point is that your example should demonstrate what you consider the best (or just "good") use of fixtures. Not some use case that doesn't appear useful.

Reply:
Maybe the fixures help with reuse? Since they advertise what object the test wants, the test framework can provide a pre-existing object.

But this should be written in C++, not macros.

Reply:
> In general, I always ask that people who add test-framework improvements do this together with an actual use for these improvements. In this example, I don't see anything in the post-fixture version of the example you used that was better than we had before (at least, I didn't understand where it is better). Can you perhaps use a fixture in a different test which make it more obvious how wonderful this feature is and how much better/shorter/clearer/safer/whatever it makes tests?

i explained in previous comment that 

> the reason to change tls_test.cc is just to demonstrate how SEASTAR_THREAD_FIXURE_TEST_CASE works. but it might not be the best place for this. i can drop this change if it hurts the readability in this very case.

i haven't found a test in Seastar which replicate the patter of https://github.com/scylladb/scylladb/blob/23be6f03360ba4943da55ac5128c7f6398264476/test/boost/sstable_compaction_test.cc#L939-L1007, probably i didn't try hard enough.

> 
> From my experience with pytest, the most important feature of fixtures is the ability to share them between multiple tests - e.g., set up a table once and then let 10 different test functions use the same table. This allows the developer to create many very small tests, without paying overhead (like creating the same table again and again 10 times). ‎The result is that instead of having huge test functions that test 20 different things (like happens often in dtest), you can write very small test functions that each tests just one small feature - and when you have a failure you can see immediately where it failed. If I understand correctly, the fixture feature you added here - in the way you added it - doesn't support this ability. These fixtures get created and destroyed separately for each test. My question then is why are they any better than RAII, which already allows C++ code to create something in a beginning of a test and ensure it gets destroyed when the test ends no matter how it gets destroyed.

you are talking about a fixture with session or suite scope. this is not the pain point i am trying to address in this change.

Reply:
> Maybe the fixures help with reuse? Since they advertise what object the test wants, the test framework can provide a pre-existing object.

exactly. to be more specific, by "reuse",  we are not reusing the same object across different test cases. but to reuse the code.

Reply:
> you are talking about a fixture with session or suite scope. this is not the pain point i am trying to address in this change.

I see. So it's still not clear to me what pain point you are solving that RAII didn't. But in any case, since you're inventing a new fixture API, and since shared ("session scope") fixtures are part of the fixture feature in every test framework I know, shouldn't you at least think ahead how your API will fit that case as well?

And I agree with Avi, let's try to think before we invent new macros whether the same thing can be implemented using C++ features such as RII, inheritance or templates.

Reply:
> > Maybe the fixures help with reuse? Since they advertise what object the test wants, the test framework can provide a pre-existing object.
> 
> exactly. to be more specific, by "reuse", we are not reusing the same object across different test cases. but to reuse the code.

But to reuse code, why isn't it enough to have global functions (not lambdas) or classes with methods, and so on? I still don't understand the use case, I'm afraid.

Reply:
yeah, i agree. the global functions and helper classes can also achieve this goal. and fixture, probably, should provide data or runtime facilities, i.e., fixtures, not the code which manipulates them.

i've been using fixtures for both purposes in previous projects. so i wanted to keep using fixtures in scylladb the same way. alas, probably, i should go with a helper class/function as you suggest. and leave C++ implementation for another day, when we really need the fixture support in Seastar.

Reply:
> I see. So it's still not clear to me what pain point you are solving that RAII didn't. But in any case, since you're inventing a new fixture API, and since shared ("session scope") fixtures are part of the fixture feature in every test framework I know, shouldn't you at least think ahead how your API will fit that case as well?

i did think about this. i even read related code and document in Boost.Test. but i didn't dump what i had thought or researched in the commit message or the code if it does not help. i don't really want to make the cover letter looks like the first part of a thesis. it is not closely related to the PR and it does not contribute to the design decision. sorry for repeating my comment, i just didn't think it's in the scope of this PR.